In [16]:
import pandas as pd 

In [57]:
import pandas as pd
import logging
import os

# Log fayl joylashuvi
log_path = r"C:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\Log\data_loader.log"

# Log konfiguratsiyasi
os.makedirs(os.path.dirname(log_path), exist_ok=True)
logging.basicConfig(
    filename=log_path,
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
    filemode="a"  # append mode
)

try:
    logging.info("Data o'qish jarayoni boshlandi.")
    df = pd.read_csv(r"C:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\Data\Raw_Data\global_ai_ml_data_salaries.csv")
    logging.info(f"Data muvaffaqiyatli o'qildi: {df.shape} o'lchamda.")
except FileNotFoundError as e:
    logging.error(f"Fayl topilmadi: {e}")
except Exception as e:
    logging.error(f"Noma’lum xatolik yuz berdi: {e}")


In [18]:
df.head()

work_year  experience_level  employment_type  job_title  salary  \
0       2024                 3                2          0  296300   
1       2024                 3                2          0  166600   
2       2024                 3                2          0  264200   
3       2024                 3                2          0  143100   
4       2024                 3                2          0   50000   

   salary_currency  salary_in_usd  employee_residence  remote_ratio  \
0                9         296300                  33             0   
1                9         166600                  33             0   
2                9         264200                  33             0   
3                9         143100                  33             0   
4                3          55555                  22             0   

   company_location  company_size  salary_per_hour  is_remote  \
0                32             1       142.451923          0   
1                32             1        80.096154          0   
2                32             1       127.019231          0   
3                32             1        68.798077          0   
4                22             1        26.709135          0   

   is_large_company  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4144 entries, 0 to 4143
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   work_year           4144 non-null   int64  
 1   experience_level    4144 non-null   int64  
 2   employment_type     4144 non-null   int64  
 3   job_title           4144 non-null   int64  
 4   salary              4144 non-null   int64  
 5   salary_currency     4144 non-null   int64  
 6   salary_in_usd       4144 non-null   int64  
 7   employee_residence  4144 non-null   int64  
 8   remote_ratio        4144 non-null   int64  
 9   company_location    4144 non-null   int64  
 10  company_size        4144 non-null   int64  
 11  salary_per_hour     4144 non-null   float64
 12  is_remote           4144 non-null   int64  
 13  is_large_company    4144 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 453.4 KB


In [1]:
import sys
import os

In [2]:
sys.path.append(os.path.abspath("../Source"))

In [19]:
import sys
print(sys.executable)


c:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\myvenv\Scripts\python.exe


In [20]:
from preprocessing import Cleaner, Encoder, Scaler

In [22]:
cleaner = Cleaner(df)
df = cleaner.tozala().get_df()

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40332 entries, 0 to 40331
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           40332 non-null  int64 
 1   experience_level    40332 non-null  object
 2   employment_type     40332 non-null  object
 3   job_title           40332 non-null  object
 4   salary              40332 non-null  int64 
 5   salary_currency     40332 non-null  object
 6   salary_in_usd       40332 non-null  int64 
 7   employee_residence  40332 non-null  object
 8   remote_ratio        40332 non-null  int64 
 9   company_location    40332 non-null  object
 10  company_size        40332 non-null  object
dtypes: int64(4), object(7)
memory usage: 3.4+ MB


In [26]:

encoder = Encoder(df)
df = encoder.encodla().get_df()

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40332 entries, 0 to 40331
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   work_year            40332 non-null  int64
 1   job_title            40332 non-null  int64
 2   salary               40332 non-null  int64
 3   salary_currency      40332 non-null  int64
 4   salary_in_usd        40332 non-null  int64
 5   employee_residence   40332 non-null  int64
 6   remote_ratio         40332 non-null  int64
 7   company_location     40332 non-null  int64
 8   experience_level_EN  40332 non-null  int64
 9   experience_level_EX  40332 non-null  int64
 10  experience_level_MI  40332 non-null  int64
 11  experience_level_SE  40332 non-null  int64
 12  employment_type_CT   40332 non-null  int64
 13  employment_type_FL   40332 non-null  int64
 14  employment_type_FT   40332 non-null  int64
 15  employment_type_PT   40332 non-null  int64
 16  company_size_L       4

In [30]:
scaler = Scaler(df)
df = scaler.scaling_qil().get_df()

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40332 entries, 0 to 40331
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   work_year            40332 non-null  float64
 1   job_title            40332 non-null  float64
 2   salary               40332 non-null  float64
 3   salary_currency      40332 non-null  float64
 4   salary_in_usd        40332 non-null  float64
 5   employee_residence   40332 non-null  float64
 6   remote_ratio         40332 non-null  float64
 7   company_location     40332 non-null  float64
 8   experience_level_EN  40332 non-null  float64
 9   experience_level_EX  40332 non-null  float64
 10  experience_level_MI  40332 non-null  float64
 11  experience_level_SE  40332 non-null  float64
 12  employment_type_CT   40332 non-null  float64
 13  employment_type_FL   40332 non-null  float64
 14  employment_type_FT   40332 non-null  float64
 15  employment_type_PT   40332 non-null 

In [32]:
df.head()

work_year  job_title    salary  salary_currency  salary_in_usd  \
0   0.533424  -0.719710  0.018352         0.220287       0.140113   
1   0.533424  -0.719710 -0.159937         0.220287      -0.418953   
2   0.533424   0.575812 -0.327013         0.220287      -0.942856   
3   0.533424   0.575812 -0.478321         0.220287      -1.417317   
4   0.533424   0.661044  0.152644         0.220287       0.561217   

   employee_residence  remote_ratio  company_location  experience_level_EN  \
0            0.304715     -0.542407          0.303171            -0.307153   
1            0.304715     -0.542407          0.303171            -0.307153   
2            0.304715      1.860795          0.303171            -0.307153   
3            0.304715      1.860795          0.303171            -0.307153   
4            0.304715      1.860795          0.303171            -0.307153   

   experience_level_EX  experience_level_MI  experience_level_SE  \
0            -0.150043            -0.640438             0.814945   
1            -0.150043            -0.640438             0.814945   
2            -0.150043             1.561432            -1.227076   
3            -0.150043             1.561432            -1.227076   
4            -0.150043            -0.640438             0.814945   

   employment_type_CT  employment_type_FL  employment_type_FT  \
0           -0.036953           -0.018634            0.056646   
1           -0.036953           -0.018634            0.056646   
2           -0.036953           -0.018634            0.056646   
3           -0.036953           -0.018634            0.056646   
4           -0.036953           -0.018634            0.056646   

   employment_type_PT  company_size_L  company_size_M  company_size_S  
0           -0.038599       -0.197964        0.211123       -0.070594  
1           -0.038599       -0.197964        0.211123       -0.070594  
2           -0.038599       -0.197964        0.211123       -0.070594  
3           -0.038599       -0.197964        0.211123       -0.070594  
4           -0.038599       -0.197964        0.211123       -0.070594

In [34]:
df.to_csv(r"C:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\Data\Preprocessing_Data\Data_Preprocessing.csv", index=False)


# ML Engineers Dataset

In [38]:
df = pd.read_csv(r"C:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\Data\Feature_Selection/Feature_Enginering.csv")

# Umumiy nechta NaN borligini ko‘rish

In [48]:
print("Umumiy NaN qiymatlar soni:", df.isna().sum().sum())


Umumiy NaN qiymatlar soni: 0


# Har bir ustun bo‘yicha NaN lar soni

In [49]:
print(df.isna().sum())


work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
salary_per_hour       0
is_remote             0
is_large_company      0
dtype: int64


In [40]:
df.head()

work_year  experience_level  employment_type  job_title  salary  \
0       2024                 3                2          0  296300   
1       2024                 3                2          0  166600   
2       2024                 3                2          0  264200   
3       2024                 3                2          0  143100   
4       2024                 3                2          0   50000   

   salary_currency  salary_in_usd  employee_residence  remote_ratio  \
0                9         296300                  33             0   
1                9         166600                  33             0   
2                9         264200                  33             0   
3                9         143100                  33             0   
4                3          55555                  22             0   

   company_location  company_size  salary_per_hour  is_remote  \
0                32             1       142.451923          0   
1                32             1        80.096154          0   
2                32             1       127.019231          0   
3                32             1        68.798077          0   
4                22             1        26.709135          0   

   is_large_company  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4144 entries, 0 to 4143
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   work_year           4144 non-null   int64  
 1   experience_level    4144 non-null   int64  
 2   employment_type     4144 non-null   int64  
 3   job_title           4144 non-null   int64  
 4   salary              4144 non-null   int64  
 5   salary_currency     4144 non-null   int64  
 6   salary_in_usd       4144 non-null   int64  
 7   employee_residence  4144 non-null   int64  
 8   remote_ratio        4144 non-null   int64  
 9   company_location    4144 non-null   int64  
 10  company_size        4144 non-null   int64  
 11  salary_per_hour     4144 non-null   float64
 12  is_remote           4144 non-null   int64  
 13  is_large_company    4144 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 453.4 KB


# Kategorik ustunlarni Label Encoding qilish

In [43]:
from sklearn.preprocessing import LabelEncoder  # sklearn’dan LabelEncoder chaqiramiz

# DataFrame’dagi faqat matnli (object) ustunlarni aniqlaymiz
cat_cols = df.select_dtypes(include="object").columns

# Har bir ustun uchun encoderlarni keyin ishlatish uchun saqlash
label_encoders = {}

# Kategorik ustunlarni sonlarga aylantiramiz
for col in cat_cols:
    le = LabelEncoder()         # Encoder obyektini yaratamiz
    df[col] = le.fit_transform(df[col])  # Har bir kategoriyani raqamga o‘tkazamiz
    label_encoders[col] = le    # Keyin kerak bo‘lsa qayta ishlatish uchun saqlaymiz

print("Kategorik ustunlar kodlandi ✅")


Kategorik ustunlar kodlandi ✅


In [44]:
df.head()

work_year  experience_level  employment_type  job_title  salary  \
0       2024                 3                2          0  296300   
1       2024                 3                2          0  166600   
2       2024                 3                2          0  264200   
3       2024                 3                2          0  143100   
4       2024                 3                2          0   50000   

   salary_currency  salary_in_usd  employee_residence  remote_ratio  \
0                9         296300                  33             0   
1                9         166600                  33             0   
2                9         264200                  33             0   
3                9         143100                  33             0   
4                3          55555                  22             0   

   company_location  company_size  salary_per_hour  is_remote  \
0                32             1       142.451923          0   
1                32             1        80.096154          0   
2                32             1       127.019231          0   
3                32             1        68.798077          0   
4                22             1        26.709135          0   

   is_large_company  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4144 entries, 0 to 4143
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   work_year           4144 non-null   int64  
 1   experience_level    4144 non-null   int64  
 2   employment_type     4144 non-null   int64  
 3   job_title           4144 non-null   int64  
 4   salary              4144 non-null   int64  
 5   salary_currency     4144 non-null   int64  
 6   salary_in_usd       4144 non-null   int64  
 7   employee_residence  4144 non-null   int64  
 8   remote_ratio        4144 non-null   int64  
 9   company_location    4144 non-null   int64  
 10  company_size        4144 non-null   int64  
 11  salary_per_hour     4144 non-null   float64
 12  is_remote           4144 non-null   int64  
 13  is_large_company    4144 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 453.4 KB


# Sonli ustunlarni Scaling qilish

In [46]:
from sklearn.preprocessing import StandardScaler  # Normalizatsiya uchun chaqiramiz

# DataFrame’dan faqat sonli ustunlarni tanlab olamiz
num_cols = df.select_dtypes(include=["int64", "float64"]).columns

# StandardScaler obyektini yaratamiz
scaler = StandardScaler()

# Sonli ustunlarni standard ko‘rinishga keltiramiz (o‘rtacha=0, std=1)
df[num_cols] = scaler.fit_transform(df[num_cols])

print("Sonli ustunlar scale qilindi ✅")


Sonli ustunlar scale qilindi ✅


In [47]:
df.head()

work_year  experience_level  employment_type  job_title    salary  \
0   0.588637          0.490385         0.039737        0.0  0.467775   
1   0.588637          0.490385         0.039737        0.0 -0.185757   
2   0.588637          0.490385         0.039737        0.0  0.306030   
3   0.588637          0.490385         0.039737        0.0 -0.304169   
4   0.588637          0.490385         0.039737        0.0 -0.773281   

   salary_currency  salary_in_usd  employee_residence  remote_ratio  \
0         0.156696       1.382085            0.283656     -0.494994   
1         0.156696      -0.436319            0.283656     -0.494994   
2         0.156696       0.932041            0.283656     -0.494994   
3         0.156696      -0.765790            0.283656     -0.494994   
4        -6.583934      -1.993178           -1.475793     -0.494994   

   company_location  company_size  salary_per_hour  is_remote  \
0          0.282288      0.064148         1.382085  -0.488737   
1          0.282288      0.064148        -0.436319  -0.488737   
2          0.282288      0.064148         0.932041  -0.488737   
3          0.282288      0.064148        -0.765790  -0.488737   
4         -1.383623      0.064148        -1.993178  -0.488737   

   is_large_company  
0         -0.139407  
1         -0.139407  
2         -0.139407  
3         -0.139407  
4         -0.139407

In [50]:
df.to_csv(r"C:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\Data\Preprocessing_Data\ML_Preprocessing.csv", index=False)

In [51]:
# Barcha sonli ustunlar bo‘yicha skewness
skew_values = df.skew(numeric_only=True)

print("Skewness qiymatlari:")
print(skew_values)


Skewness qiymatlari:
work_year             -1.816688
experience_level      -2.362001
employment_type      -21.879590
job_title              0.000000
salary                30.776827
salary_currency       -6.486558
salary_in_usd          0.746662
employee_residence    -3.412598
remote_ratio           1.536418
company_location      -3.430186
company_size          -2.149497
salary_per_hour        0.746662
is_remote              1.557917
is_large_company       7.036403
dtype: float64


# 📊 Skewness Tahlili va Xulosa

## Ustunlar bo‘yicha tahlil

- **work_year (-1.81)** → chapga qiyshaygan.  
  ➝ Ko‘p ma’lumotlar katta yillarda, eski yillar kamroq.

- **experience_level (-2.36)** → kuchli chapga qiyshaygan.  
  ➝ Ko‘pchilik junior yoki mid-level bo‘lishi mumkin.

- **employment_type (-21.87)** → juda kuchli chapga qiyshaygan.  
  ➝ Ehtimol bitta asosiy qiymat ("FT") ko‘pchilikda. One-hot encoding yoki rare label encoding tavsiya qilinadi.

- **job_title (0.00)** → mukammal simmetrik.  
  ➝ Bu ustun aslida kategorik, son sifatida qaralmasligi kerak.

- **salary (30.77)** → juda kuchli o‘ngga qiyshaygan.  
  ➝ Katta maoshlar juda ozchilikda bor. **Log transformatsiya** tavsiya qilinadi.

- **salary_currency (-6.48)** → chapga qiyshaygan.  
  ➝ Ko‘p hollarda USD, boshqa valyutalar kam.

- **salary_in_usd (0.74)** → biroz o‘ngga qiyshaygan.  
  ➝ Normal taqsimotga yaqin.

- **employee_residence (-3.41)** → chapga qiyshaygan.  
  ➝ Ko‘pchilik xodimlar bitta mamlakatda (masalan, AQSh) yashaydi.

- **remote_ratio (1.53)** → o‘ngga qiyshaygan.  
  ➝ Ko‘pchilik remote ishlaydi.

- **company_location (-3.43)** → chapga qiyshaygan.  
  ➝ Ko‘p kompaniyalar bitta mamlakatda joylashgan.

- **company_size (-2.14)** → chapga qiyshaygan.  
  ➝ Ko‘pchilik kompaniyalar "M" yoki "S" bo‘lishi mumkin.

- **salary_per_hour (0.74)** → biroz o‘ngga qiyshaygan.  
  ➝ Normalga yaqin.

- **is_remote (1.55)** → o‘ngga qiyshaygan.  
  ➝ Ko‘pchilik remote ishlaydi.

- **is_large_company (7.03)** → kuchli o‘ngga qiyshaygan.  
  ➝ Katta kompaniyalar ozchilik, ko‘pchilik kichik/yopiq kompaniyalarda ishlaydi.

---

## 📝 Umumiy xulosa

1. **Ko‘plab ustunlar kategorik bo‘lib, son sifatida qaralgani uchun skewness yuqori**.  
2. **salary va salary_in_usd** ustunlariga **log transformatsiya** qo‘llash foydali.  
3. **employment_type, job_title, salary_currency, company_location, employee_residence, company_size** ustunlari uchun **One-Hot Encoding yoki Target Encoding** tavsiya qilinadi.  
4. **is_remote va is_large_company** ustunlari **imbalanced (nomutanosib)** bo‘lgani uchun balanslash texnikalari (**SMOTE, oversampling, undersampling**) kerak bo‘lishi mumkin.


In [53]:
import numpy as np
from sklearn.preprocessing import LabelEncoder, PowerTransformer

# ================================
# 1. Salary ustunlarini log transformatsiya qilish
# ================================
df["salary"] = np.log1p(df["salary"])                 # log(1+x) -> outlierlarni kamaytiradi
df["salary_in_usd"] = np.log1p(df["salary_in_usd"])
df["salary_per_hour"] = np.log1p(df["salary_per_hour"])

# ================================
# 2. Kategorik ustunlarni encoding qilish
# ================================
categorical_cols = ["employment_type", "job_title", "salary_currency",
                    "employee_residence", "company_location", "company_size"]

for col in categorical_cols:
    le = LabelEncoder()
    df[col] = le.fit_transform(df[col])   # LabelEncoder orqali raqamlashtiramiz

# ================================
# 3. work_year va experience_level ustunlari (Yeo-Johnson)
# ================================
pt = PowerTransformer(method="yeo-johnson")

df[["work_year", "experience_level"]] = pt.fit_transform(
    df[["work_year", "experience_level"]]
)

# ================================
# 4. is_remote, is_large_company
# ================================
# Bu ustunlar binary (0/1) bo‘lgani uchun transformatsiya qilinmaydi

# ================================
# ✅ Natijani ko‘rish
# ================================
print(df.head())


   work_year  experience_level  employment_type  job_title    salary  \
0   0.644596          0.584505                2          0  0.324796   
1   0.644596          0.584505                2          0 -0.230037   
2   0.644596          0.584505                2          0  0.236646   
3   0.644596          0.584505                2          0 -0.450433   
4   0.644596          0.584505                2          0       NaN   

   salary_currency  salary_in_usd  employee_residence  remote_ratio  \
0                9       0.624856                  33     -0.494994   
1                9      -0.851594                  33     -0.494994   
2                9       0.505960                  33     -0.494994   
3                9            NaN                  33     -0.494994   
4                3            NaN                  22     -0.494994   

   company_location  company_size  salary_per_hour  is_remote  \
0                32             1         0.624856  -0.488737   
1         

c:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\myvenv\Lib\site-packages\pandas\core\arraylike.py:399: RuntimeWarning: invalid value encountered in log1p
  result = getattr(ufunc, method)(*inputs, **kwargs)


| Feature            | Oldingi Qiymat | Keyingi Qiymat | Farq (Keyingi - Oldingi) |
|--------------------|----------------|----------------|--------------------------|
| work_year          | -1.816688      | -0.911461      | 0.905227                 |
| experience_level   | -2.362001      | -1.132064      | 1.229937                 |
| employment_type    | -21.879590     | -21.879590     | 0.000000                 |
| job_title          | 0.000000       | 0.000000       | 0.000000                 |
| salary             | 30.776827      | -2.144479      | -32.921306               |
| salary_currency    | -6.486558      | -6.486558      | 0.000000                 |
| salary_in_usd      | 0.746662       | -2.163460      | -2.910122                |
| employee_residence | -3.412598      | -3.412598      | 0.000000                 |
| remote_ratio       | 1.536418       | 1.536418       | 0.000000                 |
| company_location   | -3.430186      | -3.430186      | 0.000000                 |
| company_size       | -2.149497      | -2.149497      | 0.000000                 |
| salary_per_hour    | 0.746662       | -2.163460      | -2.910122                |
| is_remote          | 1.557917       | 1.557917       | 0.000000                 |
| is_large_company   | 7.036403       | 7.036403       | 0.000000                 |


In [56]:
df.to_csv(r"C:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\Data\Preprocessing_Data\ML_After_Skewness.csv", index=False)